### PR TITLE
Migration unreachable code

### PR DESCRIPTION
**Pull request by lindamartin**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 1]
```diff
@@ -41,35 +47,103 @@
 @ActivatedByDefault
 @SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
 @SqaleConstantRemediation("5 min")
-public class UnreachableCodeCheck extends SquidCheck<LexerlessGrammar> {
+public class UnreachableCodeCheck extends SubscriptionBaseVisitor {
+
+  private Deque<Boolean> blockLevel = new ArrayDeque<>();
+  private static Kind[] JUMP_STATEMENT = {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: MINOR Rename this field "JUMP_STATEMENT" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS3008)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 2]
```diff
@@ -41,35 +47,103 @@
...
+    Kind.BREAK_STATEMENT,
+    Kind.RETURN_STATEMENT,
+    Kind.CONTINUE_STATEMENT,
+    Kind.THROW_STATEMENT
+  };
+  private static Kind[] STATEMENTS = {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: MINOR Rename this field "STATEMENTS" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS3008)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 3]
```diff
@@ -41,35 +47,103 @@
...
-  public static boolean isUnReachableCode(AstNode node) {
-    return node != null
-      && !node.is(
-        Kind.ELSE_CLAUSE,
-        Kind.FUNCTION_DECLARATION,
-        Kind.GENERATOR_DECLARATION,
-        Kind.CLASS_DECLARATION,
-      EcmaScriptPunctuator.RCURLYBRACE)
-      && !(node.is(Kind.EXPRESSION_STATEMENT) && ((ExpressionStatementTree) node).expression().is(Kind.CLASS_EXPRESSION));
+  private boolean isExcludedExpression(Tree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: MAJOR Make "isExcludedExpression" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java [hunk 4]
```diff
@@ -41,35 +47,103 @@
...
+      getContext().addIssue(this, tree, "This statement can't be reached and so start a dead code block.");
+      updateStateTo(false);
+    }
+
+    if (tree.is(JUMP_STATEMENT)) {
+      updateStateTo(true);
+    }
+  }
+
+  private boolean isScopeWithoutBlock(Tree tree) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnreachableCodeCheck.java]: MAJOR Make "isScopeWithoutBlock" a "static" method. [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues:
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Add the "@Override" annotation above this method signature](https://github.com/SonarCommunity/sonar-javascript/blob/033b336ba11dc5257e0135dd0a7a014ff4d54f44/javascript-squid/src/main/java/org/sonar/plugins/javascript/api/tree/statement/ForOfStatementTree.java#L48) [rule](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)

**Comment by pynicolas** [comment]
Looks good to me.

**Pull request closed without merging**